payroll/models/hr_salary_rule.py

```diff
@@ -178,77 +178,84 @@ def _recursive_search_of_rules(self):
 
+    def _reset_localdict_values(self, localdict):
+        localdict["result_name"] = None
+        localdict["result_qty"] = 1.0
+        localdict["result_rate"] = 100
+        localdict["result"] = None
+        return localdict
+
     # TODO should add some checks on the type of result (should be float)
     def _compute_rule(self, localdict):
         """
         :param localdict: dictionary containing the environement in which to
                           compute the rule
-        :return: returns a tuple build as the base/amount computed, the quantity
-                 and the rate
-        :rtype: (float, float, float)
+        :return: returns a dict with values for the payslip line.
+                 The dict should minimum have "name", "quantity", "rate" and "amount".
+        :rtype: {"name": string, "quantity": float, "rate": float, "amount": float}
         """
         self.ensure_one()
-        if self.amount_select == "fix":
-            try:
-                return (
-                    self.amount_fix,
-                    float(safe_eval(self.quantity, localdict)),
-                    100.0,
-                    False,  # result_name is always False if not computed by python.
-                )
-            except Exception:
-                raise UserError(
-                    _("Wrong quantity defined for salary rule %s (%s).")
-                    % (self.name, self.code)
-                )
-        elif self.amount_select == "percentage":
-            try:
-                return (
-                    float(safe_eval(self.amount_percentage_base, localdict)),
-                    float(safe_eval(self.quantity, localdict)),
-                    self.amount_percentage,
-                    False,  # result_name is always False if not computed by python.
-                )
-            except Exception:
-                raise UserError(
-                    _(
-                        "Wrong percentage base or quantity defined for salary "
-                        "rule %s (%s)."
-                    )
-                    % (self.name, self.code)
-                )
-        else:
-            try:
-                safe_eval(
-                    self.amount_python_compute, localdict, mode="exec", nocopy=True
-                )
-                result_qty = 1.0
-                result_rate = 100.0
-                result_name = False
-                if "result_name" in localdict:
-                    result_name = localdict["result_name"]
-                if "result_qty" in localdict:
-                    result_qty = localdict["result_qty"]
-                if "result_rate" in localdict:
-                    result_rate = localdict["result_rate"]
-                return (
-                    float(localdict["result"]),
-                    float(result_qty),
-                    float(result_rate),
-                    result_name,
+        method = "_compute_rule_{}".format(self.amount_select)
```

Comment:
:+1: this is an elegant solution to the question of "how to make salary rule computation as flexible as possible as simply as possible"